File: Data Structures/Linked List/singleLinkedList.cpp
#include <iostream>
using namespace std;

class Node {                            // node structure
    public:
    int data;
    Node * link;
    Node(int d){
        data = d;
    }
};

class LinkedList {
    public:
    Node * head;
    void append(int data){
        if (head == NULL) {                     // if head is empty
            head = new Node(data);
            return;
        }                                       // if head is not empty
        Node * ptr = head;
        while (ptr->link != NULL) {
            ptr = ptr->link;
        }
        Node * fresh = new Node(data);
        ptr->link = fresh;
        fresh->link = NULL;
    }
    void preappend(int data){                   // append at the front
        if (head == NULL) {                     // if head is empty
            head = new Node(data);
            return;
        }                                       // if head is not empty
        Node * fresh = new Node(data);
        fresh->link = head;
        head = fresh;
    }
    void append_after(int data, int key){       // append after the key data
        if (head == NULL) {
            cout << "list underflow" << endl;
            return;
        }
        Node * ptr = head;
        while (ptr->link != NULL && ptr->data != key) {
            ptr = ptr->link;
        }
        Node * fresh = new Node(data);
        fresh->link = ptr->link;
        ptr->link = fresh;
    }
    void display(){                             // display the linked list
        Node * ptr = head;
        while (ptr != NULL) {
            cout << ptr->data <<" ";
            ptr = ptr->link;
        }
        cout << endl;
    }
};

int main(){
    LinkedList * head = new LinkedList();
    head->append(20);
    head->preappend(10);
    head->append(30);
    head->append_after(40, 30);
    head->display();
    return 0;
}

File: Data Structures/Linked List/singleLinkedList.py
# single linked list using python
# node structure
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

# single linked list structure
class SingleLinkedList:
    def __init__(self, node=None):
        self.head = node

    def append(self, data):                         # append node at the end of the list
        if self.head == None:
            self.head = Node(data)
        else:
            temp = self.head
            while temp.next != None:
                temp = temp.next
            temp.next = Node(data)
    
    def pop_back(self):                             # delete element at the back of the list
        if self.head == None:
            print('List has no element. Can\'t pop')
        else:
            temp = self.head
            while temp.next.next != None:
                temp = temp.next
            temp.next = None

    def display(self):                              # display the entire list
        if self.head == None:
            print('Linked list is empty')
        else:
            temp = self.head
            while temp != None:
                print(temp.data, end=' ')
                temp = temp.next
            print()

# driving code
if __name__ == '__main__':
    ll = SingleLinkedList()
    ll.append(10)
    ll.append(20)
    ll.append(30)
    ll.display()
    ll.pop_back()
    ll.display()
